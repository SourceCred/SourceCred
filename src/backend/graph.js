// @flow

export type Address = {
  repositoryName: string,
  pluginName: string,
  id: string,
};

export type Node<T> = {
  address: Address,
  inEdges: Address[],
  outEdges: Address[],
  payload: T,
};

export type Edge<T> = {
  address: Address,
  src: Address,
  dst: Address,
  payload: T,
};

export type Graph = {
  nodes: {[stringAddress: string]: Node<mixed>},
  edges: {[stringAddress: string]: Edge<mixed>},
};

export function addressToString(address: Address) {
  if (address.repositoryName.includes("$")) {
    const escaped = JSON.stringify(address.repositoryName);
    throw new Error(`address.repositoryName must not include "\$": ${escaped}`);
  }
  if (address.pluginName.includes("$")) {
    const escaped = JSON.stringify(address.pluginName);
    throw new Error(`address.pluginName must not include "\$": ${escaped}`);
  }
  if (address.id.includes("$")) {
    const escaped = JSON.stringify(address.id);
    throw new Error(`address.id must not include "\$": ${escaped}`);
  }
  return `${address.repositoryName}\$${address.pluginName}\$${address.id}`;
}

export function stringToAddress(string: string) {
  const parts = string.split("$");
  if (parts.length !== 3) {
    const escaped = JSON.stringify(string);
    throw new Error(`Input should have exactly two \$s: ${escaped}`);
  }
  return {
    repositoryName: parts[0],
    pluginName: parts[1],
    id: parts[2],
  };
}
