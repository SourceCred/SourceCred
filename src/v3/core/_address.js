// @flow
// This module implements Address functionality for the Graph module
// This module should not be directly imported by clients; rather, all public
// parts of this module are re-exported via Graph

import stringify from "json-stable-stringify";

export opaque type NodeAddress: string = string;
export opaque type EdgeAddress: string = string;
type GenericAddress = NodeAddress | EdgeAddress;

const NODE_PREFIX = "N";
const EDGE_PREFIX = "E";
const NODE_CODE_POINT = NODE_PREFIX.charCodeAt(0);
const EDGE_CODE_POINT = EDGE_PREFIX.charCodeAt(0);
const SEPARATOR = "\0";

const NODE: "NODE" = "NODE";
const EDGE: "EDGE" = "EDGE";
type AddressType = typeof NODE | typeof EDGE;

function addressType(x: string): ?AddressType {
  if (x != null && x.endsWith(SEPARATOR)) {
    if (x.charCodeAt(0) === NODE_CODE_POINT && x.charCodeAt(1) === 0) {
      return NODE;
    }
    if (x.charCodeAt(0) === EDGE_CODE_POINT && x.charCodeAt(1) === 0) {
      return EDGE;
    }
  }
}

export function assertNodeAddress(x: NodeAddress, what: string = "node") {
  const type = addressType(x);
  switch (type) {
    case "NODE":
      return;
    case "EDGE":
      throw new Error(
        `${what}: expected NodeAddress, got EdgeAddress: ${stringify(x)}`
      );
    case null:
    case undefined:
      throw new Error(`${what}: bad address: ${stringify(x)}`);
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`${what}: invariant violation: ${stringify(x)}`);
  }
}

export function assertEdgeAddress(x: EdgeAddress, what: string = "edge") {
  const type = addressType(x);
  switch (type) {
    case "NODE":
      throw new Error(
        `${what}: expected EdgeAddress, got NodeAddress: ${stringify(x)}`
      );
    case "EDGE":
      return;
    case null:
    case undefined:
      throw new Error(`${what}: bad address: ${stringify(x)}`);
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`${what}: invariant violation: ${stringify(x)}`);
  }
}

export function assertAddress(x: GenericAddress, what: string = "address") {
  if (addressType(x) == null) {
    throw new Error(
      `${what}: expected NodeAddress or EdgeAddress, got: ${stringify(x)}`
    );
  }
}

export function assertAddressArray(
  arr: $ReadOnlyArray<string>,
  what: string = "address array"
) {
  if (arr == null) {
    throw new Error(String(arr));
  }
  arr.forEach((s: string) => {
    if (s == null) {
      throw new Error(`${what}: ${String(s)} in ${stringify(arr)}`);
    }
    if (s.indexOf(SEPARATOR) !== -1) {
      throw new Error(`${what}: NUL char: ${stringify(arr)}`);
    }
  });
}

function nullDelimited(components: $ReadOnlyArray<string>): string {
  return [...components, ""].join(SEPARATOR);
}

export function nodeAddress(arr: $ReadOnlyArray<string>): NodeAddress {
  assertAddressArray(arr);
  return NODE_PREFIX + SEPARATOR + nullDelimited(arr);
}

export function edgeAddress(arr: $ReadOnlyArray<string>): EdgeAddress {
  assertAddressArray(arr);
  return EDGE_PREFIX + SEPARATOR + nullDelimited(arr);
}

export function toParts(a: GenericAddress): string[] {
  assertAddress(a);
  const parts = a.split(SEPARATOR);
  return parts.slice(1, parts.length - 1);
}

export function nodeAppend(
  base: NodeAddress,
  ...components: string[]
): NodeAddress {
  assertNodeAddress(base);
  assertAddressArray(components);
  return base + nullDelimited(components);
}

export function edgeAppend(
  base: EdgeAddress,
  ...components: string[]
): EdgeAddress {
  assertEdgeAddress(base);
  assertAddressArray(components);
  return base + nullDelimited(components);
}

export function nodeToString(a: NodeAddress): string {
  assertNodeAddress(a);
  const parts = toParts(a);
  return `nodeAddress(${stringify(parts)})`;
}

export function edgeToString(a: EdgeAddress): string {
  assertEdgeAddress(a);
  const parts = toParts(a);
  return `edgeAddress(${stringify(parts)})`;
}

/**
 * Determine whether the parts of `prefix` form a prefix of the parts of
 * `address`. That is, determine whether there exists an `i` such that
 * `toParts(prefix)` equals `toParts(address).slice(0, i)`.
 */
export function nodeHasPrefix(
  address: NodeAddress,
  prefix: NodeAddress
): boolean {
  assertNodeAddress(address);
  assertNodeAddress(prefix);
  return address.startsWith(prefix);
}

/**
 * Determine whether the parts of `prefix` form a prefix of the parts of
 * `address`. That is, determine whether there exists an `i` such that
 * `toParts(prefix)` equals `toParts(address).slice(0, i)`.
 */
export function edgeHasPrefix(
  address: EdgeAddress,
  prefix: EdgeAddress
): boolean {
  assertEdgeAddress(address);
  assertEdgeAddress(prefix);
  return address.startsWith(prefix);
}
