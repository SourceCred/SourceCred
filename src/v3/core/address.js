// @flow
// This module implements Address functionality for the Graph module
// This module should not be directly imported by clients; rather, all public
// parts of this module are re-exported via Graph

import stringify from "json-stable-stringify";

export opaque type _NodeAddress = string;
export opaque type _EdgeAddress = string;
type GenericAddress = _NodeAddress | _EdgeAddress;

export const _Address = {
  toParts,
  nodeAddress,
  edgeAddress,
};

const NODE_PREFIX = "N";
const EDGE_PREFIX = "E";
const SEPARATOR = "\0";

const NODE = Symbol("NODE");
const EDGE = Symbol("EDGE");
type AddressType = typeof NODE | typeof EDGE;

function addressType(x: string): ?AddressType {
  if (x != null && x.endsWith(SEPARATOR)) {
    if (x.startsWith(NODE_PREFIX)) {
      return NODE;
    }
    if (x.startsWith(EDGE_PREFIX)) {
      return EDGE;
    }
  }
}

export function assertNodeAddress(x: _NodeAddress) {
  const type = addressType(x);
  switch (type) {
    case NODE:
      return;
    case EDGE:
      throw new Error(`Expected NodeAddress, got EdgeAddress: ${x}`);
    default:
      throw new Error(`Bad address: ${x}`);
  }
}

export function assertEdgeAddress(x: _EdgeAddress) {
  const type = addressType(x);
  switch (type) {
    case NODE:
      throw new Error(`Expected EdgeAddress, got NodeAddress: ${x}`);
    case EDGE:
      return;
    default:
      throw new Error(`Bad address: ${x}`);
  }
}
function assertAddress(x: GenericAddress) {
  if (addressType(x) == null) {
    throw new Error(`Expected Address, got: ${x}`);
  }
}

function assertAddressArray(arr: $ReadOnlyArray<string>) {
  if (arr == null) {
    throw new Error(String(arr));
  }
  arr.forEach((s: string) => {
    if (s == null) {
      throw new Error(`${String(s)} in ${stringify(arr)}`);
    }
    if (s.indexOf(SEPARATOR) !== -1) {
      throw new Error(`NUL char: ${stringify(arr)}`);
    }
  });
}

function nodeAddress(arr: $ReadOnlyArray<string> | void): _NodeAddress {
  if (arr === undefined) {
    return NODE_PREFIX + SEPARATOR;
  } else {
    assertAddressArray(arr);
    return [NODE_PREFIX, ...arr, ""].join(SEPARATOR);
  }
}

function edgeAddress(arr: $ReadOnlyArray<string> | void): _EdgeAddress {
  if (arr === undefined) {
    return EDGE_PREFIX + SEPARATOR;
  } else {
    assertAddressArray(arr);
    return [EDGE_PREFIX, ...arr, ""].join(SEPARATOR);
  }
}

function toParts(a: GenericAddress): string[] {
  assertAddress(a);
  const parts = a.split(SEPARATOR);
  return parts.slice(1, parts.length - 1);
}
