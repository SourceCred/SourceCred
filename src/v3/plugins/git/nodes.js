// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const BLOB_TYPE: "BLOB" = "BLOB";
export const COMMIT_TYPE: "COMMIT" = "COMMIT";
export const SUBMODULE_COMMIT_TYPE: "SUBMODULE_COMMIT" = "SUBMODULE_COMMIT";
export const TREE_TYPE: "TREE" = "TREE";
export const TREE_ENTRY_TYPE: "TREE_ENTRY" = "TREE_ENTRY";

export const _Prefix = Object.freeze({
  base: GIT_PREFIX,
  blob: _gitAddress(BLOB_TYPE),
  commit: _gitAddress(COMMIT_TYPE),
  submoduleCommit: _gitAddress(SUBMODULE_COMMIT_TYPE),
  tree: _gitAddress(TREE_TYPE),
  treeEntry: _gitAddress(TREE_ENTRY_TYPE),
});

export type BlobAddress = {|
  +type: typeof BLOB_TYPE,
  +hash: Hash,
|};
export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};
export type SubmoduleCommitAddress = {|
  +type: typeof SUBMODULE_COMMIT_TYPE,
  +submoduleUrl: string,
  +commitHash: Hash,
|};
export type TreeAddress = {|
  +type: typeof TREE_TYPE,
  +hash: Hash,
|};
export type TreeEntryAddress = {|
  +type: typeof TREE_ENTRY_TYPE,
  +treeHash: Hash,
  +name: string,
|};

// A tree entry has contents with one of the following types of
// addresses.
export type TreeEntryContentsAddress =
  | BlobAddress
  | TreeAddress
  | SubmoduleCommitAddress;

export type StructuredAddress =
  | BlobAddress
  | CommitAddress
  | SubmoduleCommitAddress
  | TreeAddress
  | TreeEntryAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "BLOB": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: BLOB_TYPE, hash};
    }
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    case "SUBMODULE_COMMIT": {
      if (rest.length !== 2) throw fail();
      const [submoduleUrl, commitHash] = rest;
      return {type: SUBMODULE_COMMIT_TYPE, submoduleUrl, commitHash};
    }
    case "TREE": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: TREE_TYPE, hash};
    }
    case "TREE_ENTRY": {
      if (rest.length !== 2) throw fail();
      const [treeHash, name] = rest;
      return {type: TREE_ENTRY_TYPE, treeHash, name};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case BLOB_TYPE:
      return NodeAddress.append(_Prefix.blob, x.hash);
    case COMMIT_TYPE:
      return NodeAddress.append(_Prefix.commit, x.hash);
    case SUBMODULE_COMMIT_TYPE:
      return NodeAddress.append(
        _Prefix.submoduleCommit,
        x.submoduleUrl,
        x.commitHash
      );
    case TREE_TYPE:
      return NodeAddress.append(_Prefix.tree, x.hash);
    case TREE_ENTRY_TYPE:
      return NodeAddress.append(_Prefix.treeEntry, x.treeHash, x.name);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
