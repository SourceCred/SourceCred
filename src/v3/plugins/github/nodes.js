// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
function githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export type RepoAddress = {|
  +type: "REPO",
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: "ISSUE",
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: "PULL",
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: "REVIEW",
  +pull: PullAddress,
  +fragment: string,
|};
export type CommentAddress = {|
  +type: "COMMENT",
  +parent: IssueAddress | PullAddress | ReviewAddress,
  +fragment: string,
|};
export type UserlikeAddress = {|
  +type: "USERLIKE",
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case "repo": {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: "REPO", owner, name};
    }
    case "issue": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "ISSUE", repo, number};
    }
    case "pull": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "PULL", repo, number};
    }
    case "review": {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, fragment] = rest;
      const repo = {type: "REPO", owner, name};
      const pull = {type: "PULL", repo, number: pullNumber};
      return {type: "REVIEW", pull, fragment};
    }
    case "comment": {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case "issue": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const issue = {type: "ISSUE", repo, number: issueNumber};
          return {type: "COMMENT", parent: issue, fragment};
        }
        case "pull": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          return {type: "COMMENT", parent: pull, fragment};
        }
        case "review": {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          const review = {type: "REVIEW", pull, fragment: reviewFragment};
          return {type: "COMMENT", parent: review, fragment};
        }
        default:
          throw fail();
      }
    }
    case "userlike": {
      if (rest.length !== 1) {
        throw fail();
      }
      const [login] = rest;
      return {type: "USERLIKE", login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case "REPO":
      return githubAddress("repo", x.owner, x.name);
    case "ISSUE":
      return githubAddress("issue", x.repo.owner, x.repo.name, x.number);
    case "PULL":
      return githubAddress("pull", x.repo.owner, x.repo.name, x.number);
    case "REVIEW":
      return githubAddress(
        "review",
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.fragment
      );
    case "COMMENT":
      switch (x.parent.type) {
        case "ISSUE":
          return githubAddress(
            "comment",
            "issue",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "PULL":
          return githubAddress(
            "comment",
            "pull",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "REVIEW":
          return githubAddress(
            "comment",
            "review",
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.fragment,
            x.fragment
          );
        default:
          // eslint-disable-next-line no-unused-expressions
          (x.parent.type: empty);
          throw new Error(`Bad comment parent type: ${x.parent.type}`);
      }
    case "USERLIKE":
      return githubAddress("userlike", x.login);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
