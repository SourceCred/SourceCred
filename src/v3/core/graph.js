// @flow

import type {_NodeAddress, _EdgeAddress} from "./address";
import {_Address} from "./address";

export const Address = _Address;
export type NodeAddress = _NodeAddress;
export type EdgeAddress = _EdgeAddress;
export type Edge = {|
  +address: EdgeAddress,
  +src: NodeAddress,
  +dst: NodeAddress,
|};

export type Neighbor = {|+node: NodeAddress, +edge: Edge|};
export opaque type Direction = Symbol;
export const IN: Direction = Symbol("IN");
export const OUT: Direction = Symbol("OUT");
export type NeighborsOptions = {|
  +direction: ?Direction,
  +nodePrefix: ?NodeAddress,
  +edgePrefix: ?EdgeAddress,
|};

export opaque type GraphJSON = any; // TODO

export class Graph {
  _nodes: Set<NodeAddress>;
  // If `e` is an Edge in the graph, then:
  // * _edges.get(e.address) `deepEquals` e
  // * _inEdges.get(e.dst) `contains` e
  // * _outEdges.get(e.src) `contains` e
  _edges: Map<EdgeAddress, Edge>;
  _inEdges: Map<NodeAddress, Edge[]>;
  _outEdges: Map<NodeAddress, Edge[]>;

  constructor(): void {
    this._nodes = new Set();
    this._edges = new Map();
    this._inEdges = new Map();
    this._outEdges = new Map();
  }

  addNode(a: NodeAddress): this {
    const _ = a;
    throw new Error("addNode");
  }

  removeNode(a: NodeAddress): this {
    const _ = a;
    throw new Error("removeNode");
  }

  hasNode(a: NodeAddress): boolean {
    const _ = a;
    throw new Error("hasNode");
  }

  nodes(): Iterator<NodeAddress> {
    throw new Error("nodes");
  }

  addEdge({src, dst, address}: Edge): this {
    const _ = {src, dst, address};
    throw new Error("addEdge");
  }

  removeEdge(a: EdgeAddress): this {
    const _ = a;
    throw new Error("removeEdge");
  }

  hasEdge(address: EdgeAddress): boolean {
    const _ = address;
    throw new Error("hasEdge");
  }

  edge(address: EdgeAddress): ?Edge {
    const _ = address;
    throw new Error("edge");
  }

  edges(): Iterator<Edge> {
    throw new Error("edges");
  }

  neighbors(node: NodeAddress, options?: NeighborsOptions): Iterator<Neighbor> {
    const _ = {node, options};
    throw new Error("neighbors");
  }

  copy(): Graph {
    throw new Error("copy");
  }

  toJSON(): GraphJSON {
    throw new Error("toJSON");
  }

  static fromJSON(json: GraphJSON): Graph {
    const _ = json;
    throw new Error("fromJSON");
  }

  static merge(graphs: Iterable<Graph>): Graph {
    const _ = graphs;
    throw new Error("merge");
  }
}
