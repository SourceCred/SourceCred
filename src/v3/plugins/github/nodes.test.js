// @flow

import {NodeAddress} from "../../core/graph";
import {fromRaw, toRaw} from "./nodes";

describe("plugins/github/nodes", () => {
  const repo = () => ({
    type: "REPO",
    owner: "sourcecred",
    name: "example-github",
  });
  const issue = () => ({type: "ISSUE", repo: repo(), number: "2"});
  const pull = () => ({type: "PULL", repo: repo(), number: "5"});
  const review = () => ({type: "REVIEW", pull: pull(), fragment: "100313899"});
  const issueComment = () => ({
    type: "COMMENT",
    parent: issue(),
    fragment: "373768703",
  });
  const pullComment = () => ({
    type: "COMMENT",
    parent: pull(),
    fragment: "396430464",
  });
  const reviewComment = () => ({
    type: "COMMENT",
    parent: review(),
    fragment: "171460198",
  });
  const user = () => ({type: "USERLIKE", login: "decentralion"});
  const examples = {
    repo,
    issue,
    pull,
    review,
    issueComment,
    pullComment,
    reviewComment,
    user,
  };

  describe("`fromRaw` after `toRaw` is identity", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        expect(fromRaw(toRaw(instance))).toEqual(instance);
      });
    });
  });

  describe("`toRaw` after `fromRaw` is identity", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        const raw = toRaw(instance);
        expect(toRaw(fromRaw(raw))).toEqual(raw);
      });
    });
  });

  describe("snapshots as expected:", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        const raw = NodeAddress.toParts(toRaw(instance));
        expect({address: raw, structured: instance}).toMatchSnapshot();
      });
    });
  });

  describe("errors on", () => {
    describe("fromRaw(...) with", () => {
      function expectBadAddress(name: string, parts: $ReadOnlyArray<string>) {
        it(name, () => {
          const address = NodeAddress.fromParts([
            "sourcecred",
            "github",
            ...parts,
          ]);
          // $ExpectFlowError
          expect(() => fromRaw(address)).toThrow("Bad address");
        });
      }
      function checkBadCases(
        partses: $ReadOnlyArray<{|
          +name: string,
          +parts: $ReadOnlyArray<string>,
        |}>
      ) {
        let partsAccumulator = [];
        for (const {name, parts} of partses) {
          const theseParts = [...partsAccumulator, ...parts];
          expectBadAddress(name, theseParts);
          partsAccumulator = theseParts;
        }
      }
      it("undefined", () => {
        // $ExpectFlowError
        expect(() => fromRaw(undefined)).toThrow("undefined");
      });
      it("null", () => {
        // $ExpectFlowError
        expect(() => fromRaw(null)).toThrow("null");
      });
      it("with bad prefix", () => {
        // $ExpectFlowError
        expect(() => fromRaw(NodeAddress.fromParts(["foo"]))).toThrow(
          "Bad address"
        );
      });
      expectBadAddress("no kind", []);
      describe("repository with", () => {
        checkBadCases([
          {name: "no owner", parts: ["repo"]},
          {name: "no name", parts: ["owner"]},
          {name: "extra parts", parts: ["name", "foo"]},
        ]);
      });
      describe("issue with", () => {
        checkBadCases([
          {name: "no owner", parts: ["issue"]},
          {name: "no name", parts: ["owner"]},
          {name: "no number", parts: ["name"]},
          {name: "extra parts", parts: ["123", "foo"]},
        ]);
      });
      describe("pull request with", () => {
        checkBadCases([
          {name: "no owner", parts: ["pull"]},
          {name: "no name", parts: ["owner"]},
          {name: "no number", parts: ["name"]},
          {name: "extra parts", parts: ["123", "foo"]},
        ]);
      });
      describe("pull request review with", () => {
        checkBadCases([
          {name: "no owner", parts: ["review"]},
          {name: "no name", parts: ["owner"]},
          {name: "no number", parts: ["name"]},
          {name: "no fragment", parts: ["123"]},
          {name: "extra parts", parts: ["987", "foo"]},
        ]);
      });
      describe("comment", () => {
        expectBadAddress("with no subkind", ["comment"]);
        expectBadAddress("with bad subkind", ["comment", "icecream"]);
        describe("on issue with", () => {
          checkBadCases([
            {name: "no owner", parts: ["comment", "issue"]},
            {name: "no name", parts: ["owner"]},
            {name: "no number", parts: ["name"]},
            {name: "no fragment", parts: ["123"]},
            {name: "extra parts", parts: ["987", "foo"]},
          ]);
        });
        describe("on pull request with", () => {
          checkBadCases([
            {name: "no owner", parts: ["comment", "pull"]},
            {name: "no name", parts: ["owner"]},
            {name: "no number", parts: ["name"]},
            {name: "no fragment", parts: ["123"]},
            {name: "extra parts", parts: ["987", "foo"]},
          ]);
        });
        describe("on pull request review with", () => {
          checkBadCases([
            {name: "no owner", parts: ["comment", "review"]},
            {name: "no name", parts: ["owner"]},
            {name: "no number", parts: ["name"]},
            {name: "no review fragment", parts: ["123"]},
            {name: "no comment fragment", parts: ["987"]},
            {name: "extra parts", parts: ["654", "foo"]},
          ]);
        });
      });
      describe("userlike", () => {
        checkBadCases([
          {name: "no login", parts: ["userlike"]},
          {name: "extra parts", parts: ["decentra", "lion"]},
        ]);
      });
    });

    describe("toRaw(...) with", () => {
      it("null", () => {
        // $ExpectFlowError
        expect(() => toRaw(null)).toThrow("null");
      });
      it("undefined", () => {
        // $ExpectFlowError
        expect(() => toRaw(undefined)).toThrow("undefined");
      });
      it("bad type", () => {
        // $ExpectFlowError
        expect(() => toRaw({type: "ICE_CREAM"})).toThrow("Unexpected type");
      });
      it("bad comment type", () => {
        expect(() => {
          // $ExpectFlowError
          toRaw({type: "COMMENT", parent: {type: "ICE_CREAM"}});
        }).toThrow("Bad comment parent type");
      });
    });
  });
});
