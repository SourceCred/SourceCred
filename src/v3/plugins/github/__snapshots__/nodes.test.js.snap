// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/nodes snapshots as expected: issue 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "issue",
    "sourcecred",
    "example-github",
    "2",
  ],
  "structured": Object {
    "number": "2",
    "repo": Object {
      "name": "example-github",
      "owner": "sourcecred",
      "type": "REPO",
    },
    "type": "ISSUE",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: issueComment 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "comment",
    "issue",
    "sourcecred",
    "example-github",
    "2",
    "373768703",
  ],
  "structured": Object {
    "fragment": "373768703",
    "parent": Object {
      "number": "2",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "type": "COMMENT",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: pull 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "pull",
    "sourcecred",
    "example-github",
    "5",
  ],
  "structured": Object {
    "number": "5",
    "repo": Object {
      "name": "example-github",
      "owner": "sourcecred",
      "type": "REPO",
    },
    "type": "PULL",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: pullComment 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "comment",
    "pull",
    "sourcecred",
    "example-github",
    "5",
    "396430464",
  ],
  "structured": Object {
    "fragment": "396430464",
    "parent": Object {
      "number": "5",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
    "type": "COMMENT",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: repo 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "repo",
    "sourcecred",
    "example-github",
  ],
  "structured": Object {
    "name": "example-github",
    "owner": "sourcecred",
    "type": "REPO",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: review 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "review",
    "sourcecred",
    "example-github",
    "5",
    "100313899",
  ],
  "structured": Object {
    "fragment": "100313899",
    "pull": Object {
      "number": "5",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
    "type": "REVIEW",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: reviewComment 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "comment",
    "review",
    "sourcecred",
    "example-github",
    "5",
    "100313899",
    "171460198",
  ],
  "structured": Object {
    "fragment": "171460198",
    "parent": Object {
      "fragment": "100313899",
      "pull": Object {
        "number": "5",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "PULL",
      },
      "type": "REVIEW",
    },
    "type": "COMMENT",
  },
}
`;

exports[`plugins/github/nodes snapshots as expected: user 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "github",
    "userlike",
    "decentralion",
  ],
  "structured": Object {
    "login": "decentralion",
    "type": "USERLIKE",
  },
}
`;
